subprojects/version-control/src/integTest/groovy/org/gradle/vcs/git/internal/VcsCleanupIntegrationTest.groovy

```diff
@@ -0,0 +1,197 @@
+        commits["latest.integration"] = repo.commit("version is snapshot", repo.workTree)
+
+        buildFile << """
+            apply plugin: 'base'
+            
+            configurations {
+                conf
+            }
+            dependencies {
+                conf "org.test:dep:" + repoVersion
+            }
+            task assertVersion {
+                dependsOn configurations.conf
+                doLast {
+                    def files = zipTree(configurations.conf.singleFile).files
+                    def versionFile = files.find { it.name == "version" }
+                    assert versionFile
+                    assert versionFile.text == repoVersion
+                }
+            }
+        """
+        settingsFile << """
+            sourceControl {
+                vcsMappings {
+                    withModule("org.test:dep") {
+                        from vcs(GitVersionControlSpec) {
+                            url = "${repo.url}"
+                        }
+                    }
+                }
+            }
+        """
+    }
+
+    def "does not remove vcs checkout on every build"() {
+        succeeds("assertVersion", "-PrepoVersion=1.0")
+        def checkout = checkoutDir("dep", commits.initial.id.name, repo.id)
+
+        // Put new file in the checkout directory that would be deleted if Gradle
+        // deletes the checkout directory
+        def trashFile = checkout.file("trash")
+        trashFile.text = "junk"
+
+        when:
+        succeeds("assertVersion", "-PrepoVersion=1.0")
+        then:
+        trashFile.assertExists()
+
+        when:
+        cleanupNow()
+        succeeds("assertVersion", "-PrepoVersion=1.0")
+        then:
+        // checkout was used recently, so it should still be kept around
+        trashFile.assertExists()
+    }
+
+    def "removes vcs checkout after 7 days"() {
+        // checkout all versions
+        versions.each { version ->
+            succeeds("assertVersion", "-PrepoVersion=${version}")
+        }
+
+        // Mark 1.0 as unused
+        markUnused("1.0")
+
+        when:
+        cleanupNow()
+        succeeds("assertVersion", "-PrepoVersion=3.0")
+        then:
+        checkoutDir("dep", commits["1.0"].id.name, repo.id).assertDoesNotExist()
+        checkoutDir("dep", commits["2.0"].id.name, repo.id).assertExists()
+        checkoutDir("dep", commits["3.0"].id.name, repo.id).assertExists()
+    }
+
+    def "does not remove vcs checkout that is older than 7 days but recently used"() {
+        // checkout all versions
+        versions.each { version ->
+            succeeds("assertVersion", "-PrepoVersion=${version}")
+        }
+        // mark all versions as unused
+        versions.each { version ->
+            markUnused(version)
+        }
+
+        when:
+        cleanupNow()
+        succeeds("assertVersion", "-PrepoVersion=1.0")
+        then:
+        checkoutDir("dep", commits["1.0"].id.name, repo.id).assertExists()
```

Comment:
I don't think we can't tell the difference here between "was not removed by cleanup" and "was removed by clean up and then checked out again". Maybe we don't care at this point.

Reply:
I think we don't care. Cleanup happens after any checkout can happen.  It occurs when we close the `PersistentCache`.